packages/subscribable/src/async-iterable.ts

```diff
@@ -0,0 +1,172 @@
+                    data: TData;
+                }
+              | {
+                    __type: PublishType.ERROR;
+                    err: unknown;
+                }
+          )[];
+      }
+    | {
+          __hasPolled: true;
+          onData: (data: TData) => void;
+          onError: Parameters<ConstructorParameters<typeof Promise>[0]>[1];
+      };
+
+let EXPLICIT_ABORT_TOKEN: symbol;
+function createExplicitAbortToken() {
+    // This function is an annoying workaround to prevent `process.env.NODE_ENV` from appearing at
+    // the top level of this module and thwarting an optimizing compiler's attempt to tree-shake.
+    return Symbol(
+        __DEV__
+            ? "This symbol is thrown from a socket's iterator when the connection is explicitly " +
+                  'aborted by the user'
+            : undefined,
+    );
+}
+
+const UNINITIALIZED = Symbol();
+
+export function createAsyncIterableFromDataPublisher<TData>({
+    abortSignal,
+    dataChannelName,
+    dataPublisher,
+    errorChannelName,
+}: Config): AsyncIterable<TData> {
+    const iteratorState: Map<IteratorKey, IteratorState<TData>> = new Map();
+    function publishErrorToAllIterators(reason: unknown) {
+        for (const [iteratorKey, state] of iteratorState.entries()) {
+            if (state.__hasPolled) {
+                iteratorState.delete(iteratorKey);
+                state.onError(reason);
+            } else {
+                state.publishQueue.push({
+                    __type: PublishType.ERROR,
+                    err: reason,
+                });
+            }
+        }
+    }
+    const abortController = new AbortController();
+    abortSignal.addEventListener('abort', () => {
+        abortController.abort();
+        publishErrorToAllIterators((EXPLICIT_ABORT_TOKEN ||= createExplicitAbortToken()));
+    });
+    const options = { signal: abortController.signal } as const;
+    let firstError: unknown | typeof UNINITIALIZED = UNINITIALIZED;
+    dataPublisher.on(
+        errorChannelName,
+        err => {
+            if (firstError === UNINITIALIZED) {
+                firstError = err;
+                abortController.abort();
+                publishErrorToAllIterators(err);
+            }
+        },
+        options,
+    );
+    dataPublisher.on(
+        dataChannelName,
+        data => {
+            iteratorState.forEach((state, iteratorKey) => {
+                if (state.__hasPolled) {
+                    const { onData } = state;
+                    iteratorState.set(iteratorKey, { __hasPolled: false, publishQueue: [] });
+                    onData(data as TData);
+                } else {
+                    state.publishQueue.push({
+                        __type: PublishType.DATA,
+                        data: data as TData,
+                    });
+                }
+            });
+        },
+        options,
+    );
+    return {
+        async *[Symbol.asyncIterator]() {
```

Comment:
This implementation still uses generators instead of the manual approach cancellation (via `return()`) in #3172. This means that the only way to interrupt this generator is to call the abort signal, which aborts the whole subscription – you can't just call `.return()` on the iterator. This is _probably_ sufficient for most use cases today, but we should consider rolling in some of the ideas and tests from #3172 anyway.